pkg/didcomm/messenger/messenger.go

```diff
@@ -77,16 +81,84 @@ func (m *Messenger) HandleInbound(msg service.DIDCommMsgMap, myDID, theirDID str
 		return fmt.Errorf("threadID: %w", err)
 	}
 
+	if err := m.withMetadata(thID, msg); err != nil {
+		return fmt.Errorf("with metadata: %w", err)
+	}
+
 	// saves message payload
 	return m.saveRecord(msg.ID(), record{MyDID: myDID, TheirDID: theirDID, ThreadID: thID})
 }
 
+func (m *Messenger) saveMetadata(msg service.DIDCommMsgMap) error {
+	if msg[jsonThread] == nil {
+		return nil
+	}
+
+	thread, ok := msg[jsonThread].(map[string]interface{})
+	if !ok || thread == nil {
+		return nil
+	}
+
+	metadata, ok := thread[jsonMetadata].(map[string]interface{})
+	if !ok || metadata == nil {
+		return nil
+	}
+
+	thID, err := msg.ThreadID()
+	if err != nil {
+		return fmt.Errorf("threadID: %w", err)
+	}
+
+	delete(thread, jsonMetadata)
+
+	// if the thread had only metadata we remove it from the message as well
+	if len(thread) == 0 {
+		delete(msg, jsonThread)
+	}
+
+	return m.saveRecord(fmt.Sprintf(metadataKey, thID), record{Metadata: metadata})
+}
+
+func (m *Messenger) withMetadata(thID string, msg service.DIDCommMsgMap) error {
```

Comment:
I'd prefer to use the prefix "with" to indicate an Option.

Could we rename this one?

Reply:
Sure

Reply:
I've changed it to `populateMetadata`